src/arithmetic/arithmetic_expression.c

```diff
@@ -226,19 +227,23 @@ SIValue AR_EXP_Evaluate(AR_ExpNode *root, const Record r) {
 			SIValue sub_trees[root->op.child_count];
 			for(int child_idx = 0; child_idx < root->op.child_count; child_idx++) {
 				SIValue v = AR_EXP_Evaluate(root->op.children[child_idx], r);
-				if(SIValue_IsError(v)) return v;
+				if(QueryCtx_EncounteredError()) return v;
 				sub_trees[child_idx] = v;
 			}
 
 			/* Validate before evaluation. */
 			char *error;
 			if(!_AR_EXP_ValidateInvocation(root->op.f, sub_trees, root->op.child_count, &error)) {
-				// Clean up.
+				// The expression tree failed its validations and prepared an error message.
+				// Free all associated memory.
 				for(int child_idx = 0; child_idx < root->op.child_count; child_idx++) {
 					SIValue_Free(sub_trees + child_idx);
 				}
-				QueryCtx_SetError(error);
-				jmp_buf *env = QueryCtx_GetEnv();
+				QueryCtx_SetError(error); // Set the query-level error.
+
+				// Invoke the exception handler, exiting this routine and returning to
+				// the point on the stack where the handler was instantiated.
+				jmp_buf *env = QueryCtx_GetExceptionHandler();
```

Comment:
This seems a bit too low level, consider introducing a single function which performs this logic.